aio/tools/transforms/angular-base-package/post-processors/auto-link-code.js

```diff
@@ -30,51 +30,64 @@ module.exports = function autoLinkCode(getDocFromAlias) {
-                parent.children.splice(index, 1, createLinkNode(docs[0], node.value));
-              } else {
-                // Parse the text for words that we can convert to links
-                const nodes =
-                    textContent(node)
-                        .split(/([A-Za-z0-9_.-]+)/)
-                        .filter(word => word.length)
-                        .map((word, index, words) => {
-                          // remove docs that fail the custom filter tests
-                          const filteredDocs = autoLinkCodeImpl.customFilters.reduce(
-                              (docs, filter) => filter(docs, words, index), getDocFromAlias(word));
-                          return foundValidDoc(filteredDocs, word, file) ?
-                              // Create a link wrapping the text node.
-                              createLinkNode(filteredDocs[0], word) :
-                              // this is just text so push a new text node
-                              {type: 'text', value: word};
-                        });
+        visit(node, 'text', (node, ancestors) => {
+          // Only interested in text nodes that are not inside links
+          const isInsideLink = ancestors.every(ancestor => !is(ancestor, 'a'));
+          if (!isInsideLink) {
+            return;
+          }
 
-                // Replace the text node with the links and leftover text nodes
-                Array.prototype.splice.apply(parent.children, [index, 1].concat(nodes));
-              }
-            }
-          });
-        }
+          const parent = ancestors[ancestors.length - 1];
+          const index = parent.children.indexOf(node);
+
+          // Can we convert the whole text node into a doc link?
+          const docs = getDocFromAlias(node.value);
+          if (foundValidDoc(docs, node.value, file)) {
+            parent.children.splice(index, 1, createLinkNode(docs[0], node.value));
+          } else {
+            // Parse the text for words that we can convert to links
+            const nodes = getNodes(node, file);
+            // Replace the text node with the links and leftover text nodes
+            Array.prototype.splice.apply(parent.children, [index, 1].concat(nodes));
+          }
+        });
       });
     };
   }
 
+  function isValidCodeElement(node, ancestors) {
+    // Only interested in code elements that:
+    // * do not have `no-auto-link` class
+    // * do not have an ignored language
+    // * are not inside links
+    const containsElement = autoLinkCodeImpl.codeElements.some(elementType => is(node, elementType));
+    const hasNoAutoLink = !node.properties.className || !node.properties.className.includes('no-auto-link');
+    const isLanguageSupported = !autoLinkCodeImpl.ignoredLanguages.includes(node.properties.language);
+    const isInsideLink = ancestors.every(ancestor => !is(ancestor, 'a'));
+    return containsElement && hasNoAutoLink && isLanguageSupported && isInsideLink;
```

Comment:
```suggestion
    const isCodeElement = autoLinkCodeImpl.codeElements.some(elementType => is(node, elementType));
    const hasNoAutoLink = node.properties.className && node.properties.className.includes('no-auto-link');
    const isLanguageSupported = !autoLinkCodeImpl.ignoredLanguages.includes(node.properties.language);
    const isInsideLink = ancestors.some(ancestor => is(ancestor, 'a'));
    return isCodeElement && !hasNoAutoLink && isLanguageSupported && !isInsideLink;
```

BTW, one problem (which is not terribly important here, I guess) with splitting this up into intermediate variables is that we lose the shortcircuiting capabilities of `&&`/`||`. I.e. now we will always traverse all ancestors even if this is not a code element.

Can you check whether this makes any difference in the time it takes to generate the docs?

Reply:
When I ran the script I did not notice any significant delays. I could try to revert the code and use `console.time` to see how long it takes to generate the docs. Would that be sufficient?

Reply:
We could always exit early? 

```typescript
const isCodeElement = ...
if (!isCodeElement) {
  return false;
}
const hasNoAutoLink = ...
```

Reply:
Let's see if this introduces any noticeable delay first.
Any way of timing it witl and without the change is fine.

Reply:
There is no significant delay. Before the change, it takes `63.63s` to run the whole process. After the change, it takes `51.22s.` I have run it a few times and don't see much difference. 

I am printing the times that the entire process takes. It does that automatically.

Reply:
OK, if it doesn't make any difference, let's keep it as is.
It is strange though that it takes less time with the change. Where does the speed-up come from?

Reply:
Not entirely sure TBH. I see that the times are very random. Sometimes it takes `40`s, and sometimes `60`s. Either way, I do not see much difference, after running it a few times.

Reply:
:+1: